pkg/ruler/ruler.go

```diff
@@ -731,6 +740,8 @@ func (r *Ruler) listRulesShuffleSharding(ctx context.Context) (map[string]rulesp
 	mu := sync.Mutex{}
 	owned := map[string]rulespb.RuleGroupList{}
 	backedUp := map[string]rulespb.RuleGroupList{}
+	gLock := sync.Mutex{}
```

Comment:
by now you must realize that I don't like mutexes 😅  

I think you can create a slice of maps of size `concurrency` and give one dedicated map per goroutine and merge all maps from the slice after the wait group. The same approach could be used with `owned` and `backedUp` to get rid of the other mutex.

Reply:
I think the code will be much more complex is that approach. The use of mutex here seems to be acceptable considering the small scope of the lock. Maybe @yeya24 you can share your opinion on this matter.

Reply:
Not a blocker for me, but a nice to have, therefore I will approve.